---
title: "DE Analysis of Broadfield Bulk RNAseq data (DEN)"
author: "Theo Killian"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
---

```{r, echo = FALSE, results = 'asis'}
suppressPackageStartupMessages(library("dplyr"))
knitr::opts_chunk$set(collapse=TRUE, comment="#>", warning=FALSE, message=FALSE)
```

## RNAseq Analysis (DEN)

This report describes an RNAseq analysis of the bulk RNAseq data provided by
Lindsay Broadfield using tools from the
[DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)
Bioconductor package. Specifically, this pipeline analyzes the inferred
transcriptional changes between test groups and produces various diagnostic
plots. At the end of the analysis, the results of the analysis will be output as
*.csv* files listing genes that are differentially expressed and sorted by 
significantly small p-adjusted values across groups of samples. Additionally,
list enriched gene of significantly expressed genes are generated to be used for
further analyses such as
[GO enrichment](http://geneontology.org/docs/go-enrichment-analysis/). 
 
*This report compares samples on the DEN condition*

The data was processed using a standard RNAseq pipeline including
[FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
for the quality control of the raw data,
[trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic)
to trim the adapter from the sequences,
[hisat2](https://ccb.jhu.edu/software/hisat2/index.shtml)
to align the short reads to the genomic reference and
[htseq-count](https://htseq.readthedocs.io/en/release_0.11.1/count.html)
to produce the count matrix.

```{r load_libraries, warning=FALSE, message=FALSE}
library("DESeq2")
library("tidyverse")
library("biomaRt")
library("ggrepel")
library("DT")
library("gridExtra")
library("janitor")
library("openxlsx")
```

Experiment Description: <br/>
- Mice were injected with DEN or PBS at 2wks age <br/>
- Mice were randomized to HFD or remain on CD at 6 weeks age <br/>
- Mice sacrificed after 12wks HFD, liver tissue collected <br/>

For this project, there are 4 experimental groups, forming a 2x2 experimental
design, with 4 samples in each group. The 2 variables are: <br/>
- *diet (control vs high-fat diet)* <br/>
- *carcinogen injection (PBS vehicle control vs DEN carcinogen)* <br/>
 
The research questions are: <br/>
- Can any changes in metabolic pathways (glucose and glycolysis are
of particular interest) or other major cancer signaling pathways
(PI3K/Akt/mTOR signaling, Ras, Erk, etc.) be observed? <br/>
- Is there a significant amount of variation caused by the DEN injections? <br/>

At this time point in the experiment there is no significant DEN effect on most
outcomes and measurements, so we often pool the PBS and DEN samples to look for
a general diet effect. However, if there is a significant DEN effect, obviously
this approach should be reconsidered.

```{r load_data0, warning=FALSE, message=FALSE, echo=FALSE}
## load coldata
coldata <- read.xlsx("../data/12wkLiver_RNASeq_metadata.xlsx")
names(coldata) <- janitor::make_clean_names(names(coldata))
coldata$mouse_id[1] <- 2.3
coldata %>% mutate(mouse_id = as.character(mouse_id)) %>% 
  mutate(den = as.factor(case_when(
    den == 0 ~ "ctrl",
    den == 1 ~ "den"))) -> coldata

## add file names
c("rSMF095", "rSMF097", "rSMF102", "rSMF103", "rSMF089", "rSMF090", "rSMF093",
  "rSMF094", "rSMF091", "rSMF098", "rSMF100", "rSMF101", "rSMF088", "rSMF092",
  "rSMF096", "rSMF099") -> coldata$file
# head(coldata)
```

```{r load_data, warning=FALSE, message=FALSE}
## load countdata
countdata <- read.csv("../data/Broadfield_202009.csv", header = TRUE, row.names = 1)
rownames(countdata) <- countdata$gene
countdata <- countdata %>% dplyr::select(everything(), -gene)
countdata %>%
  dplyr::select(rSMF095, rSMF097, rSMF102, rSMF103, rSMF089, rSMF090, rSMF093,
                rSMF094, rSMF091, rSMF098, rSMF100, rSMF101, rSMF088, rSMF092,
                rSMF096, rSMF099) -> countdata
head(countdata)

## set analysis parameters
## define log2 fold change threshold
l2FC <- 1

## define p-adjusted value threshold
thresh_padj <- 0.05

## define organism
organism <- "Mm"
kegg_organism <- "mmu"
```

## Differential expression analysis

The standard differential expression analysis steps are wrapped into a single
function, `DESeq`. The estimation steps performed by this function are described
in the
[DESeq2 vignette](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#theory),
and in the manual page for `?DESeq` and in the Methods section of the DESeq2
publication (Love, Huber, and Anders 2014).

```{r dds, warning=FALSE, message=FALSE}
dds <- DESeqDataSetFromMatrix(countData = countdata,
                              colData = coldata,
                              design = ~ den)
cat("Number of genes in dds object *before* filtering out rows with zero counts \n")
dim(assay(dds))[1]
```

Genes with counts of all 0 for all samples are removed before running `DESeq`

```{r keep, warning=FALSE, message=FALSE}
keep <- rowSums(counts(dds)) > 0
dds <- DESeq(dds[keep, ])
cat("Number of genes in dds object *after* filtering out rows with zero counts \n")
dim(assay(dds))[1]
```

Results tables are generated using the function `results`, which extracts a
results table with log2 fold changes, p values and adjusted p values. With no
additional arguments to results, the log2 fold change and Wald test p value will
be for the last variable in the design formula, and if this is a factor, the
comparison will be the last level of this variable over the reference level.

```{r comparisons, warning=FALSE, message=FALSE}
res1 <- results(dds, contrast = c("den", "den", "ctrl"), alpha = 0.05)
```

### Experimental design

A table matching samples with the expermental conditions is shown below:

```{r coldata, warning=FALSE, message=FALSE}
knitr::kable(colData(dds))
```

### Data exploration and quality assessment (QC)

#### Principal components analysis (PCA)

PCA is a method of visually identifying the similarity or difference between
samples. PCA rotates the data cloud onto an orthagonal basis determined by the
dimensions of maximal variance. The first two Principal Components (PCs) usually
hold the majority of the variance of the data. The following plot shows the
count matrix samples projected onto the two largest Principal Components
(PC1 and PC2). 

```{r PCA1, warning=FALSE, message=FALSE}
## perform regularized log transformation and PCA and plot with ggplot2
rld <- rlogTransformation(dds)
PCA1 <- plotPCA(rld, intgroup = "group") +
        theme(plot.title = element_text(hjust = 0.5)) +
        geom_text_repel(aes(label = colnames(rld)),
                        arrow = arrow(length = unit(0.03, "npc"),
                        type = "closed", ends = "first"), force = 5) +
        ggtitle("Principal component analysis")
pdf(paste0("rna_pca.pdf"), width = 6, height = 6)
PCA1
dev.off()
```

#### Size factors 

Size factors are a method of normalizing used by the *DESeq* function to
normalize the data in terms of sequencing depth. Size factor is the median ratio
of the sample over a pseudosample: for each gene, the geometric mean of all
samples. Size factors account for differences in sequencing depth are typically
centered around 1 (indicating comparable sequencing depth).

```{r size_factor, warning=FALSE, message=FALSE}
barplot(dds$sizeFactor, main = "Size Factors for samples", col = "lavender",
        cex.axis = 0.7, las = 2, ylab = "Size Factors", cex.names = 0.9)
```

#### Dispersion plot

[DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)
estimates gene dispersion using an algorithm that first generates gene-wise 
maximum likelihood estimates (MLEs) that are obtained using only the respective
gene's data (black dots). Then, a curve (red) is fit to the MLEs to capture the
overall trend of dispersion-mean dependence. This fit is used as a prior mean
for a second estimation round, which results in the final maximum a priori (MAP)
estimates of dispersion. This results in a "shrinkage" of the noisy gene-wise
estimates toward the consensus represented by the red line. The black points
circled in blue are detected as dispersion outliers and not shrunk toward the
prior (shrinkage would follow the dotted line). A more in-depth theoretical
explanation of the DESeq2 algorithm can be found [here](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#theory)

```{r dispersion_plot, warning=FALSE, message=FALSE}
plotDispEsts(dds, main = "DESeq2 Dispersion plot")
```

### Quality assessment (QC)

#### Independent Filtering

*DESeq2* performs independent filtering by default using the mean of normalized
counts as a filter statistic. A threshold on the filter statistic (first value)
is found which optimizes the number of adjusted p-values lower than significance
level alpha. The adjusted p-values for the genes which do not pass the filter
threshold are set to NA. The results also returns the mean of normalized counts
(second value).

```{r filterThreshold}
cat("Filter thresh. val. and mean of norm. counts \n")
metadata(res1)$filterThreshold
```

#### Plot of sample rejections vs filter quantiles

The *filterThreshold* returns the threshold chosen (vertical line in the plots
below) by the *DESeq2* analysis of the lowest quantile of the filter for which
the number of sample rejections is within 1 residual standard deviation to the
peak of a curve fit to the number of rejections over the filter quantiles. The
following diagnostic plot shows the number of rejected samples (y-axis) plotted
against quantiles of filter (x-axis). 

```{r rejections_vs_quantiles1}
par(mfrow = c(1, 1))
plot(metadata(res1)$filterNumRej, type = "b", xlab = "Quantiles of filter",
     ylab = "Number of rejections", main = "DEN vs No-DEN")
lines(metadata(res1)$lo.fit, col = "red")
abline(v = metadata(res1)$filterTheta)
```

#### Histogram of frequency of p-values of results

The following plot shows the number of frequency of counts (y-axis) against 
p-values between 0 and 1 (x-axis).

```{r p-value_frequency}
par(mfrow = c(1, 1))
hist(res1$pvalue, col = "lavender", main = "DEN vs No-DEN",
     xlab = "p-values")
```

### Results

#### Summary of results of comparison of DEN vs no-DEN cells

```{r summary_res1, warning=FALSE, message=FALSE}
print(summary(res1))
```

We will match the common gene names with Ensemble ID by accessing [BioMart](https://bioconductor.org/packages/release/bioc/html/biomaRt.html).
However, to save time, the *Biomart* query is read from as saved query file.

```{r biomart, warning=FALSE, message=FALSE}
## NOTE! uncomment the following code if it is necessary to regenerate the
## ensembl_2_geneName query object.
# library("biomaRt")
# useDataset("mmusculus_gene_ensembl", useMart("ensembl")) -> mart
# getBM(attributes = c("ensembl_gene_id", "external_gene_name"),
#       mart = mart) -> mouse_ensembl_2_geneName
# names(mouse_ensembl_2_geneName) <- c("GENEID", "geneName")
# save(mouse_ensembl_2_geneName, file = "./data/mouse_ensembl_2_geneName.rda")
load(file = "../data/mouse_ensembl_2_geneName.rda")
mouse_ensembl_2_geneName %>%
  dplyr::select(GENEID, geneName) -> ensembl_2_geneName
```

Ensemble ID is matched to the common gene name via the
[BioMart](https://bioconductor.org/packages/release/bioc/html/biomaRt.html)
query and the normalized counts from the
[DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)
analysis are merged alongside as a single file.

```{r id_match, warning=FALSE, message=FALSE}
merge(as.data.frame(res1), as.data.frame(counts(dds, normalized = TRUE)),
      by = "row.names", sort = FALSE) -> resdata1
names(resdata1)[1] <- "Gene"
idx <- match(resdata1$Gene, ensembl_2_geneName$GENEID)
resdata1$geneName <- ensembl_2_geneName$geneName[idx]
resdata1 <- resdata1 %>% dplyr::select(Gene, geneName, everything()) %>%
                         dplyr::arrange(padj)
```

### MA plots

A MA plot illustrates log-fold expression change between two groups of samples,
created by transforming and the data onto two scales: *M* (the log of the ratio
of level counts for each gene between two samples) and *A* (the average level
counts for each gene across the two samples) scales. MA plots demonstrates the
difference between samples in terms of signal intensities of read counts. In
this type of plot, genes with similar expression levels in two samples will
appear around the horizontal line y = 0 (red line). The following MA plot
illustrates log-fold expression change for each comparison after the *DESeq2*
analysis. Significant genes (P < 0.05) are highlighted in red.

#### MA plot for DEN vs no-DEN cells

```{r MA1, warning=FALSE, message=FALSE}
results1 <- as.data.frame(res1)
results1[is.na(results1)] <- 0.99 # change NA results to 0.99 for correct MAplot
MA1 <- ggplot(results1, aes(x = baseMean, y = log2FoldChange)) +
       geom_point(aes(colour = padj < thresh_padj), size = 0.5) +
       scale_colour_manual(name = 'padj < 0.05',
                           values = setNames(c('red','black'), c(TRUE, FALSE))) +
       scale_x_continuous(trans = "log10", limits = c(0.1, 300000)) +
       geom_smooth(colour = "red") +
       geom_abline(slope = 0, intercept = 0, colour = "blue") +
       theme(plot.title = element_text(hjust = 0.5))
MA1
# DESeq2::plotMA(res1)
```

### Volcano plots

A volcano plot is a type of scatterplot that is used to quickly identify genes
that display large magnitude changes that are also statistically significant. A
volcano plot is constructed by plotting the negative log of the p-value on the
y-axis. This results in data points with low p-values appearing toward the top
of the plot. The x-axis is the log of the fold change between the two
experimental conditions. The log of the fold change is used so that changes in
both directions appear equidistant from the center. Plotting points in this way
results in two regions of interest in the plot: those points that are found
toward the top of the plot that are far to either the left- or right-hand sides.
These represent values that display large magnitude fold changes (on the left or
right of center) as well as high statistical significance (toward the top). The
following Volcano plot shows log of the fold change and negative log of the
p-values for each comparison. Significant genes (P < 0.05) with log2 fold change
(> 1) are highlighted in red.

#### Volcano plot for DEN vs no-DEN cells

```{r volcano1, warning=FALSE, message=FALSE}
res_final1 <- resdata1[!is.na(resdata1$padj), ]
res_final1$threshold <- as.factor(abs(res_final1$log2FoldChange) > l2FC &
                                      res_final1$padj < thresh_padj)

Vol1 <- ggplot(data = res_final1,
               aes(x = log2FoldChange, y = -log10(padj), color = threshold)) +
               geom_text(label=res_final1$geneName, nudge_x = 0.5,
                         nudge_y = 0.5, check_overlap = TRUE) +
               # geom_text_repel(data = subset(res_final1, group == "important"),
               #                 aes(label = geneName, nudge_x = -1)) +
               theme(legend.position = "none") +
               geom_point(alpha = 0.4, size = 0.5) +
               xlab("log2 fold change") + ylab("-log10 padj-value") +
               theme(plot.title = element_text(hjust = 0.5))
Vol1 + scale_color_manual(values = c("#000000", "#FF0000"))
```

### Results tables

Below are three html tables displaying the results the DE comparisons. These
tables are interactive and be queried for specific proteins or sorted by column.

Note on p-values set to NA: some values in the results table can be set to NA
for one of the following reasons:
* If within a row, all samples have zero counts, the baseMean column will be
zero, and the log2 fold change estimates, p value and adjusted p value will all
be set to NA.
* If a row contains a sample with an extreme count outlier then the p value and
adjusted p value will be set to NA. These outlier counts are detected by Cook's
distance.
* If a row is filtered by automatic independent filtering, for having a low mean
normalized count, then only the adjusted p value will be set to NA.

#### Results table for comparison of DEN vs no-DEN cells

```{r results_table1, warning=FALSE, message=FALSE}
resdata1_tab <- resdata1 %>%
                mutate_at(3:6, round, 3) %>%
                mutate_at(9:24, round, 3)
DT::datatable(resdata1_tab)
```

### GO Enrichment

GO Term Enrichment Analysis will be performed using `limma::goana()`.
This process requires `ENTREZ` identifiers, which are obtained by querying the
genome wide annotation for *M. musculus*, as well as p-adjusted values. The
enriched gene set is subjected to a hypergeometric test for differential
enrichment.

```{r warning=FALSE, message=FALSE}
library("limma")
library("GO.db")
library("org.Mm.eg.db")
resdata1$Gene <- as.character(resdata1$Gene)
resdata1 %>%
  mutate(ENTREZID = mapIds(org.Mm.eg.db, Gene, "ENTREZID", "ENSEMBL") %>%
  unname()) -> resdata1
## we need to save the results with ENTREZ IDs from now on
```

```{r warning=FALSE, message=FALSE}
## get all GO IDs, and all ENTREZ IDs associated with each ID
go_list <- mapIds(org.Mm.eg.db, keys(org.Mm.eg.db, "GO"),
                  "ENTREZID", "GO", multiVals = "list")
go_vector <- lapply(go_list, as.vector)
ezs <- sapply(go_vector, paste0, collapse = ";")
go_df <- data.frame(GOID = names(go_vector), ENTREZID = ezs)
```

Below shows the number of genes found in the original data.

```{r}
cat("Dimensions of DEN vs no-DEN cells dataset ID\n")
dim(resdata1)[1]
```

To perform GSEA, a *universe* of genes must be constructed. Each gene in the
*universe* must have an `ENTREZ ID` to be testable for differential enrichment.
Below shows the length of the list of such genes in the *universe*.

```{r warning=FALSE, message=FALSE}
resdata1_entrez <- resdata1 %>% filter(!is.na(ENTREZID))
cat("Dimensions of DEN vs no-DEN cells with an ENTREZ ID\n")
dim(resdata1_entrez)[1]
```

For reference, shown below is the the subset of the data that *does not have* an
`ENTREZ ID` and is not used to construct the universe.

```{r warning=FALSE, message=FALSE}
resdata1_no_entrez <- resdata1 %>% filter(is.na(ENTREZID))
cat("Dimensions of DEN vs no-DEN cells without an ENTREZ ID\n")
dim(resdata1_no_entrez)[1]
```

`goana` compares the *universe* of genes by looking at all GO terms for "Mm"
(*mouse*) with the gene set have provided from the `DESeq2` analysis, returning
a tibble of the GO terms that are significantly differentially enriched after
performing a hypergeometric test.

The vector of enriched genes that will be passed are compared with the universe
are those from the `DESeq2` analysis with a p-adjusted value *lower than 0.05*
and an absolute log fold change of *greater than 1*.

*We will perform GO enrichment for all genes with log2FoldChange > 1 and < -1*

```{r warning=FALSE, message=FALSE}
go_all <- goana(resdata1_entrez$ENTREZID[(resdata1_entrez$padj < thresh_padj &
                                  abs(resdata1_entrez$log2FoldChange) > l2FC)],
            resdata1_entrez$ENTREZID, organism) %>% as_tibble()

go_0 <- go_all %>% mutate(GOID = mapIds(GO.db, .$Term, "GOID", "TERM") %>%
                 unname()) %>% dplyr::select(GOID, everything()) %>%
                 arrange(P.DE) %>%
                 mutate(ADJ.P.DE = p.adjust(P.DE, method = "BH"))

# go_up <- goana(resdata1_entrez$ENTREZID[(resdata1_entrez$padj < thresh_padj &
#                                      resdata1_entrez$log2FoldChange > 1)],
#                resdata1_entrez$ENTREZID, organism) %>% as_tibble()
#
# go_1 <- go_up %>% mutate(GOID = mapIds(GO.db, .$Term, "GOID", "TERM") %>%
#                  unname()) %>% dplyr::select(GOID, everything()) %>%
#                  arrange(P.DE) %>%
#                  mutate(ADJ.P.DE = p.adjust(P.DE, method = "BH"))
#
# go_down <- goana(resdata1_entrez$ENTREZID[(resdata1_entrez$padj < thresh_padj &
#                                        resdata1_entrez$log2FoldChange < -1)],
#                  resdata1_entrez$ENTREZID, organism) %>% as_tibble()
#
# go_2 <- go_down %>% mutate(GOID = mapIds(GO.db, .$Term, "GOID", "TERM") %>%
#                    unname()) %>% dplyr::select(GOID, everything()) %>%
#                    arrange(P.DE) %>%
#                    mutate(ADJ.P.DE = p.adjust(P.DE, method = "BH"))

### then map GO terms to genes between tables via ENTREZID
go_0 <- go_0 %>% dplyr::left_join(go_df, by ="GOID")
# go_1 <- go_1 %>% dplyr::left_join(go_df, by ="GOID")
# go_2 <- go_2 %>% dplyr::left_join(go_df, by ="GOID")
```

#### GO Term Enrichment results (lfc > abs(+/-1))

Below is a html table displaying the results of significant genes found in the
GO Term Enrichment analysis for comparison of DEN vs no-DEN cells with lfc > 1

```{r}
DT::datatable(go_0)
```

### KEGG Enrichment

To do KEGG analysis, download the KEGG pathway annotation for mouse:

```{r}
## get all KEGG PathwayIDs, and all ENTREZ IDs associated with each PathwayID
GK <- getGeneKEGGLinks(species.KEGG = kegg_organism)
#saveRDS(as.data.frame(GK), file = "./app/data/kegg_path.rds")

## get all KEGG PathwayIDs, and all ENTREZ IDs associated with each PathwayID
# GK_desc <- getKEGGPathwayNames(species.KEGG = kegg_organism, remove = TRUE)

## convert all ENTREZ IDs associated with each PathwayID to a string
# GK_col <- GK %>% group_by(PathwayID) %>%
#                  summarise(GeneIDs = paste(GeneID, collapse = "|"))

# path_id_kegg <- dplyr::full_join(GK_col, GK_desc, by = "PathwayID")
# names(path_id_kegg) <- c("PathwayID", "ENTREZ_GeneIDs", "Pathway")
# saveRDS(as.data.frame(path_id_kegg), file = "./app/data/kegg_path_2_id_mouse.rds")
```

We perform the KEGG analysis in a similar manner as the GO enrichment above:

*We will perform KEGG enrichment for all genes with log2FoldChange > 1 and < -1*

```{r}
## KEGG enrichment for log2FoldChange > abs(+/-1)
kegg_0 <- kegga(resdata1_entrez$ENTREZID[(resdata1_entrez$padj < thresh_padj &
                                      abs(resdata1_entrez$log2FoldChange) > l2FC)],
                universe = resdata1_entrez$ENTREZID,
                species.KEGG = kegg_organism,
                gene.pathway = GK) %>%
                arrange(P.DE) %>%
                mutate(ADJ.P.DE = p.adjust(P.DE, method = "BH"))
```

#### KEGG Enrichment results (lfc > abs(+/-1))

```{r}
# topKEGG(kegg_0)
DT::datatable(kegg_0)
```

### Save Results

Results are saved as *.csv* and *.rds* files.

```{r write_data, warning=FALSE, message=FALSE}
# rownames(resdata1) <- NULL
write.table(resdata1, file = "./results/DEN_Broadfield_202009.csv",
            sep = ",", quote = FALSE, row.names = FALSE)
write.table(go_0, file = "./results/DEN_GO_Broadfield_202009.csv",
            sep = ",", quote = FALSE, row.names = FALSE)
write.table(kegg_0, file = "./results/DEN_KEGG_Broadfield_202009.csv",
            sep = ",", quote = FALSE, row.names = FALSE)
```

*Session Info*

```{r sessionInfo}
sessionInfo()
```

*References:*

Love, M.I., Huber, W., Anders, S. (2014) "Moderated estimation of fold change
and dispersion for RNA-seq data with DESeq2." Genome Biology, 15:550.
10.1186/s13059-014-0550-8

Anders, Simon, and Wolfgang Huber. 2010. "Differential Expression Analysis for
Sequence Count Data." Genome Biology 11:R106.
http://genomebiology.com/2010/11/10/R106.

Ritchie ME, Phipson B, Wu D, Hu Y, Law CW, Shi W, Smyth GK (2015). "limma powers
differential expression analyses for RNA-sequencing and microarray studies."
Nucleic Acids Research, 43(7), e47. doi: 10.1093/nar/gkv007. 